active/0000-stateful-helpers.md

```diff
@@ -0,0 +1,322 @@
+
+```js
+// app/helpers/greet-someone.js
+import Ember from "ember";
+
+// Usage: {{greet-someone 'bob' greeting='say hello'}}
+export default Ember.Helper.extend({
+  compute(params, hash) {
+    return `Hello ${params[0]}, nice to ${hash.greeting}`;
+  }
+});
+```
+
+When the `params` or `hash` contents change, the `compute` method is called
+again. The instance of the helper is preserved across rerenders of the parent.
+
+The `init` and `destroy` methods can be subclassed for setup and teardown.
+
+`compute` is called with the same arguments as a function helper.
+
+### Consuming an stateful helper
+
+Stateful helpers can be used anywhere an HTMLBars subexpression can be
+used. For example:
+
+```hbs
+{{#if (can-access 'admin')}}
+  {{link-to 'login'}}
+{{/if}}
+{{my-login-button isAdmin=(can-access 'admin')}}
+<my-login-button isAdmin={{can-access 'admin'}} />
+Can access? {{can-access 'admin'}}
+```
+
+In all these cases, the helper is considered one-way. That is, it is a
+readable value and not two-way bound (toggling `isAdmin` does not impact the
+helper).
+
+Let's step through exactly what happens when using an helper like this:
+
+```hbs
+<my-login-button isAdmin={{can-access 'admin'}} />
+```
+
+Upon initial render:
+
+* The helper `can-access` is looked up on the container
+* The helper is identified as a stateful helper, not a function helper
+* The helper is initialized (`init` is called)
+* The `compute` function is called on the helper.
+* The return value from `compute` is passed as an `attr` to `my-login-button`.
+* The helper instance remains in memory.
+
+If the parent scope is rerendered:
+
+* The `compute` function is called again.
+* The return value from `compute` is passed as an `attr` to `my-login-button`.
+
+Upon teardown:
+
+* The helper is destroyed, calling the `destroy` method.
+
+### Returning a value
+
+The return value of stateful and function helpers should be the return value
+of the HTMLBars subexpression they are powering. For example, given a helper:
+
+```js
+// app/helpers/full-name.js
+import Ember from "ember";
+
+// Note that what is called "options" in other docs is
+// called "blocks" here. This terminology should become the
+// documented norm.
+export default function full-name(params, hash, blocks) {
+  return params.join(' ');
+}
+```
+
+Or its equivilent stateful version, the following are effectively the same:
+
+```hbs
+<div data-name={{full-name "Fenton" "Crackshell"}}></div>
+<div data-name={{"Fenton Crackshell"}}></div>
+```
+
+```hbs
+{{my-component name=(full-name "Magica" "De Spell")}}
+{{my-component name="Magica De Spell"}}
+```
+
+```hbs
+<p>{{full-name "Bentina" "Beakley"}}</p>
+<p>{{"Bentina Beakley"}}</p>
+```
+
+An exclusion to this pattern is the following form:
+
+```hbs
+<div {{full-name "Webbigail" "Vanderquack"}}></div>
+```
+
+This is a legacy form of mustache usage. In 2.0 both forms of helpers should
+throw an exception when used in this manner.
+
+### Block form
```

Comment:
As an add on author, I agree. Subexpression use is far more important to me than block use.